### PR TITLE
Taught util.misc.mkdir() to set umask to allow group-w for makedirs c…

### PR DESCRIPTION
**Pull request by ggainey**
…alls.

Hunted down a LOT of makedirs calls and replaced with call to misc.mkdir()

fixes #7445

### CODE REVIEW

File: server/pulp/plugins/util/metadata_writer.py [hunk 1]
```diff
@@ -124,7 +124,7 @@ def _open_metadata_file_handle(self):
             parent_dir = os.path.dirname(self.metadata_file_path)
 
             if not os.path.exists(parent_dir):
```
Comment (ipanova) [inline server/pulp/plugins/util/metadata_writer.py:126]: you don't need to check for presence of the path here, remove the if condition

File: server/pulp/server/content/storage.py [hunk 2]
```diff
@@ -6,22 +6,7 @@
 from hashlib import sha256
 
 from pulp.server.config import config
-
-
-def mkdir(path):
```
Comment (ipanova) [inline server/pulp/server/content/storage.py:11]: :+1:  let's say no to the code duplication!

File: server/pulp/plugins/util/metadata_writer.py (hunk 1 above)
Comment (ggainey) [reply server/pulp/plugins/util/metadata_writer.py:126]: The os.access() check in the elif is only needed if we're not creating the path in the mkdir() call. I tried a couple of iterations here, and decided that while this one is a little redundant around the mkdir(), it seemed more clear in terms of understanding the code. What do you think?

File: server/pulp/server/async/tasks.py [hunk 3]
```diff
@@ -780,11 +782,7 @@ def _handle_cProfile(self, task_id):
         if config.getboolean('profiling', 'enabled') is True:
             self.pr.disable()
             profile_directory = config.get('profiling', 'directory')
-            try:
-                os.makedirs(profile_directory, 0755)
-            except OSError as exc:
-                if exc.errno != errno.EEXIST:
-                    raise
+            misc.mkdir(profile_directory, mode=0755)
```
Comment (ipanova) [inline server/pulp/server/async/tasks.py:785]: 0775

File: server/pulp/plugins/util/metadata_writer.py (hunk 1 above)
Comment (ipanova) [reply server/pulp/plugins/util/metadata_writer.py:126]: oooh right, i did not notice the elif statement, i am cool with leaving if

File: server/pulp/server/async/tasks.py (hunk 3 above)
Comment (ipanova) [reply server/pulp/server/async/tasks.py:785]: i think this is important and i understand correctly how umask works - if having mode 0755 the result will be:
 usr grp others 
-rwx  r-x   r-x is represented in octal as 0755
-000 000 010 umask of 002 removes any permission where there is a 1
-rwx  r-x   r-x is the result achieved requesting 0755 with umask of 002...and we are screwed :)


if having mode 0775 the result will be:
 usr grp others 
-rwx  rwx  r-x is represented in octal as 0775
-000 000 010 umask of 002 removes any permission where there is a 1
-rwx  rwx  r-x   is the result achieved requesting 0775 with umask of 002

Comment (ggainey) [reply server/pulp/server/async/tasks.py:785]: I left a few things like certificate-directories and this profiling-directory alone - they're not 'content', and are more sensitive to permissions. Unless we're migrating profiling-information, this set of perms shouldn't affect migration, and I felt it better to leave the permissions as they had been specifically set.

(Your understanding of umask is correct - it turns off permission-bits where the umask has a 1. So 0755, with the mkdirs() umask of 0002, leaves 0755)

**Review by ipanova — APPROVED** [review]
changes look good to me, i will try to test them tomorrow.

**Comment by ipanova** [comment]
@rohanpm please take a look to be sure that you are not impacted by this change in an unexpected way

File: server/pulp/plugins/util/misc.py [hunk 4]
```diff
@@ -40,13 +40,16 @@ def mkdir(*args, **kwargs):
...
     Tolerant of race conditions.
+    Sets umask to 002.
 
     :param args: path[, mode] that goes to os.makedirs
     :param kwargs: path
     :return:
     """
     try:
+        mask = os.umask(0002)
         os.makedirs(*args, **kwargs)
```
Comment (rohanpm) [inline server/pulp/plugins/util/misc.py:51]: The umask is apparently never restored if this raises an exception.  Should the original os.umask be above the "try", and the umask reset in a new "finally" block?

**Comment by rohanpm** [comment]
> @rohanpm please take a look to be sure that you are not impacted by this change in an unexpected way

Thanks Ina, I don't see any reason this would be harmful for RHSM Pulp, though I believe there is a bug where umask may not be restored when intended.

File: server/pulp/plugins/util/misc.py (hunk 4 above)
Comment (ggainey) [reply server/pulp/plugins/util/misc.py:51]: ...I am an idiot. I had exactly that at some point, and apparently lost it somewhere along the way. Thank you for the great catch, that would have been *very* embarrassing to allow out into the wild!

Comment (ggainey) [reply server/pulp/plugins/util/misc.py:51]: @rohanpm @ipanova fixed mkdir() - more eyes please, when you have a moment free

**Pull request merged**